File: temporal-io/my-python/MY/0_simple/single_node.py
"""
Example 1: Single Node Workflow
This demonstrates the simplest possible workflow with a single activity.
"""
import asyncio
from concurrent.futures import ThreadPoolExecutor
from datetime import timedelta

from temporalio import activity, workflow
from temporalio.client import Client
from temporalio.worker import Worker


# Single activity that processes data
@activity.defn
def process_data(input_value: int) -> int:
    activity.logger.info(f"Processing data: {input_value}")
    result = input_value * 2
    activity.logger.info(f"Result: {result}")
    return result


# Workflow with a single activity node
@workflow.defn
class SingleNodeWorkflow:
    @workflow.run
    async def run(self, input_value: int) -> int:
        workflow.logger.info(f"Starting workflow with input: {input_value}")

        # Execute single activity
        result = await workflow.execute_activity(
            process_data,
            input_value,
            start_to_close_timeout=timedelta(seconds=10),
        )

        workflow.logger.info(f"Workflow completed with result: {result}")
        return result


async def main():
    # Start client
    client = await Client.connect("localhost:7233")

    # Run a worker for the workflow
    async with Worker(
        client,
        task_queue="0-simple-single-node-task-queue",
        workflows=[SingleNodeWorkflow],
        activities=[process_data],
        activity_executor=ThreadPoolExecutor(5),
    ):
        # Execute the workflow
        result = await client.execute_workflow(
            SingleNodeWorkflow.run,
            42,
            id="0-simple-single-node-workflow",
            task_queue="0-simple-single-node-task-queue",
        )
        print(f"Workflow result: {result}")


if __name__ == "__main__":
    asyncio.run(main())

File: temporal-io/my-typescript/MY/0_simple/single_node.ts
/**
 * Example 1: Single Node Workflow
 * This demonstrates the simplest possible workflow with a single activity.
 *
 * This file contains:
 * - Activity definitions
 * - Workflow definitions
 * - Worker and workflow execution logic
 */

// ============================================================================
// ACTIVITIES
// ============================================================================
import { log as activityLog } from '@temporalio/activity';

/**
 * Single activity that processes data
 */
export async function processData(inputValue: number): Promise<number> {
  activityLog.info('Processing data:', { inputValue });
  const result = inputValue * 2;
  activityLog.info('Result:', { result });
  return result;
}

// ============================================================================
// WORKFLOWS
// ============================================================================
import { proxyActivities, log as workflowLog } from '@temporalio/workflow';

// Define activity interface
export interface Activities {
  processData(inputValue: number): Promise<number>;
}

// Proxy activities with timeout configuration
const { processData: processDataActivity } = proxyActivities<Activities>({
  startToCloseTimeout: '10 seconds',
});

/**
 * Simple workflow that executes a single activity
 */
export async function singleNodeWorkflow(inputValue: number): Promise<number> {
  workflowLog.info('Starting workflow with input:', { inputValue });

  // Execute single activity
  const result = await processDataActivity(inputValue);

  workflowLog.info('Workflow completed with result:', { result });
  return result;
}

// ============================================================================
// MAIN EXECUTION
// ============================================================================
import { Connection, Client } from '@temporalio/client';
import { NativeConnection, Worker } from '@temporalio/worker';
import { randomUUID } from 'crypto';

const TEMPORAL_HOST = 'localhost:7233';
const TASK_QUEUE = '0-simple-single-node-task-queue';

async function main() {
  try {
    // Connect to Temporal server
    const connection = await Connection.connect({ address: TEMPORAL_HOST });
    const client = new Client({ connection });
    console.log(`Connected to Temporal at ${TEMPORAL_HOST}`);

    // Create worker connection
    const workerConnection = await NativeConnection.connect({
      address: TEMPORAL_HOST,
    });

    // Create worker
    const worker = await Worker.create({
      connection: workerConnection,
      namespace: 'default',
      taskQueue: TASK_QUEUE,
      workflowsPath: new URL(import.meta.url).pathname,
      activities: {
        processData,
      },
    });

    console.log(`Worker started on task queue: ${TASK_QUEUE}`);

    // Run worker in background
    const workerPromise = worker.run();

    // Give worker a moment to start
    await new Promise(resolve => setTimeout(resolve, 1000));

    // Execute workflow
    const workflowId = `single-node-workflow-${randomUUID()}`;
    console.log(`Starting workflow with ID: ${workflowId}`);

    const handle = await client.workflow.start(singleNodeWorkflow, {
      args: [42],
      taskQueue: TASK_QUEUE,
      workflowId,
    });

    console.log('Workflow started, waiting for result...');
    const result = await handle.result();

    console.log(`Workflow completed successfully!`);
    console.log(`Workflow ID: ${workflowId}`);
    console.log(`Result: ${result}`);

    // Shutdown worker
    worker.shutdown();
    await workerPromise;
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

// Only run main if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}
